README.md
![VerseBot](http://i.imgur.com/zzFkW5g.png)

**A reddit bot that posts Bible verses when asked to.**

* [Usage](#usage)
  * [Making Verse Quotations](#making-verse-quotations)
  * [Editing/Deleting a VerseBot Comment](#editingdeleting-a-versebot-comment)
  * [Setting a Default Translation](#setting-a-default-translation)
* [Authors](#authors)
* [License](#license)
* [Thanks](#thanks)

## USAGE
### Making Verse Quotations
VerseBot is triggered by username mentions. Verse quotations are enclosed in brackets ([]), and may specify a desired translation within. A VerseBot quotation must follow one of the following formats (in each case the translation is optional):

```
[Book Chapter Translation]
[Book Chapter:Verse Translation]
[Book Chapter:StartVerse-EndVerse Translation]
```

More quotation formats will be implemented in the future.

### Editing/Deleting a VerseBot Comment
VerseBot comments can be easily edited or deleted by clicking on the "edit" or "delete" links that can be found at the bottom of every VerseBot comment. Only the user who triggered the VerseBot response can edit or delete the comment.

### Setting a Default Translation
A default translation is the translation VerseBot defaults to for a particular user or subreddit. This means that a user/subreddit can specify a default translation, and the user(s) do not need to place the abbreviation for the translation within their comment.

Translation defaults are prioritized in the following order:

1) User Translation

2) Subreddit Translation

3) VerseBot Translation

Default translations can be set by visiting the [Default Translations section](http://adamgrieger/versebot/#defaults) of the VerseBot website.

There are two types of default translations that can be set:

1) *User Translations* -- A set of default translations for a reddit user. Note that user default translations that have not been used for 90 days or more will be deleted to save on database storage space.

2) *Subreddit Translations* -- A set of default translations for a subreddit. This type of default translation **must** be set by a moderator of the subreddit, otherwise VerseBot will simply ignore the request. Subreddit default translations never expire.

## AUTHORS
[Matthieu Grieger](http://matthieugrieger.com)

## LICENSE
	The MIT License (MIT)

	Copyright (c) 2014, 2015 Matthieu Grieger

	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included in
	all copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
	THE SOFTWARE.

## THANKS
* [BibleGateway](http://www.biblegateway.com) and [Bible Hub](http://biblehub.com/) for being the sources of all texts posted by VerseBot.  
* [All those who contribute](https://github.com/praw-dev/praw/graphs/contributors) to [PRAW (Python Reddit API Wrapper)](https://github.com/praw-dev/praw).
* [Adam Grieger](https://github.com/adamgrieger) for the awesome VerseBot image at the top of this README.


CHANGELOG.md
# VerseBot Changelog

### January 10, 2016
* Migrated from PostgreSQL to SQLite
* Unhandled exception handling and Slack integration (implemented by [@shackra](https://github.com/shackra))
* Fixed delete and edit requests
* More PEP8 formatting and tidying up
* Changed links to VerseBot website and GitHub repository

### September 23, 2015
* Migrated to OAuth2 for authentication.
* Fixed a crash that was caused by a missing ending verse.
* Fixed crashes related to PRAW Forbidden exception handling.

### May 29, 2015
* Another large update!
* VerseBot now scans its own inbox for username mentions instead of scanning every comment on reddit.
* Added user default translation requests.
* Added subreddit default translation requests.
* New translations added to BibleGateway are automatically picked up by the bot within 24 hours.
* Fixed some reddit links that caused AutoModerator to automatically delete VerseBot comments.
* Removed Heroku-related files, not hosted on Heroku any longer.
* Added a cool new header image to the README (thanks [@adamgrieger](https://github.com/adamgrieger))!
* Replaced TaggedTanakh pickle file with Bible Hub website parsing.
* Removed NJPS translation, added JPS translation.
* Added user translation cleaning function. Deletes user translations that haven't been used for 90 days or more.
* Added edit link and delete link at bottom of every VerseBot comment which allows a user to edit/delete their own VerseBot response.

### March 27, 2015
* Added default translations for [/r/TelaIgne](https://www.reddit.com/r/TelaIgne).

### March 25, 2015
* Fixed "Back" appearing at the end of every verse.

### February 27, 2015
* Updated PRAW dependency.

### January 10, 2015
* Added default translations for [/r/DebateACatholic](https://www.reddit.com/r/DebateACatholic).

### December 17, 2014
* Added default translations for [/r/3OP](http://www.reddit.com/r/3OP).

### December 13, 2014
* Added default translations for [/r/Resurrexi](http://www.reddit.com/r/Resurrexi).

### November 15, 2014
* Added "Eph" as an abbreviation for Ephesians.

### October 7, 2014
* Reformatted code spacing and comments to comply with PEP guidelines.

### August 21, 2014
* Disabled comment edit/deletion again. Still seems to not be working. I will figure this out! Sorry.

### August 20, 2014
* Re-enabled comment edit/deletion. If you find any issues, let me know!
* Fixed some weird formatting issues in config.py.
* Removed code that incorrectly marked some edit/delete messages as read.
* Updated PRAW to 2.1.18.

### August 18, 2014
* Added NABRE translation to database.
* Set default translation for /r/Catholicism and /r/divineoffice to NABRE.

### August 9, 2014
* Temporarily disabled edit/delete actions (again) until I get the chance to fix it once and for all.

### August 6, 2014
* Fixed a bug that made edit/delete actions stop working.
* Moved comment URL in edit/delete message to comment body to avoid creating a subject over 100 characters long (thanks /u/emprags!).
* Added complete translation names and translation languages to translation statistics table.
* Improved readability of some SQL scripts.

### August 5, 2014
* Added the ability for users to edit/delete VerseBot comments. Only the user who triggers the VerseBot response will have the ability to edit/delete the comment.
* Added logic for "fixing" the the book, translation, and subreddit statistics when a VerseBot response is edited/deleted.
* Added a short sentence at the end of the comment footer that gives links for editing and deleting a comment.
* Added messages.py to implement comment editing/deletion, and more functions in the future.
* Moved all regex to regex.py to clean up the code a bit.

### July 31, 2014
* Fixed Biblical references in the book of Wisdom (deuterocanon).
* Updated createdatabase.sql to reflect above change.

### July 16, 2014
* Released the [VerseBot Statistics website!](http://matthieugrieger.com/versebot/)
* Fixed a crash that occurred when trying to reply to a recently deleted comment.
* Set /r/Lectionary default translation to NRSV.

### July 12, 2014
* Changed default translation for /r/AcademicBiblical to NRSV.

### July 9, 2014
* Fixed "Back" appearing at the end of every quotation (thanks to /u/coveredinbeeees for finding this!).

### June 30, 2014
* Compatibility changes for the new version of BibleGateway.

### June 27, 2014
* Reverted to legacy BibleGateway website for compatibility reasons. Will update to the new BibleGateway eventually.

### June 18, 2014
* Updated PRAW to address an AttributeError caused by a recent reddit update.

### June 9, 2014
* Made NRSVCE the default translation of /r/divineoffice.

### May 21, 2014
* Made KJV the default translation of /r/Protestantism.

### May 12, 2014
* Made OJB the default translation of /r/AcademicBiblical.

### May 10, 2014
* Fixed NET Bible translation. Now use the keyword 'NET' to trigger this translation.
* Fixed bot choosing wrong translations.
* Set default translation for /r/Catholicism to RSVCE.
* Added bibles/NJPS.pickle to add the NJPS translation to the bot.
* Added NJPS translation to translation stats table.
* Set default translation for /r/Judaism to NJPS.
* Implemented necessary functions and logic to incorporate pickle files once again.
* Added pickler.py. This is the script that is run to generate pickle files for Bible translations from XML.
* Fixed link to Supported Translations in VerseBot Info.md.

### May 9, 2014
* Fixed the bot mistaking books such as '1 John' as just 'John'.
* Added ordereddict to requirements.txt.

### May 8, 2014
* HUGE update!
* Completely refactored codebase.
* VerseBot now uses [BibleGateway](http://www.biblegateway.com) as its source for Bible texts. Because of this, VerseBot now supports over 100 translations in 62 languages!
* Bot automatically recognizes new translations on BibleGateway every 24 hours or so (whenever the bot is restarted automatically).

versebot/books.py
"""
VerseBot for Reddit
By Matthieu Grieger
Continued By Team VerseBot
books.py
Copyright (c) 2015 Matthieu Grieger (MIT License)
"""


def get_book(name):
    """ Retrieves a standardized book name to replace the one
    provided by the user.

    :param name: The book reference to retrieve a standardized name for
    """
    books = {
        # Old Testament
        "genesis": "Genesis",
        "gen": "Genesis",
        "gn": "Genesis",
        "bereshit": "Genesis",
        "exodus": "Exodus",
        "exod": "Exodus",
        "ex": "Exodus",
        "shemot": "Exodus",
        "leviticus": "Leviticus",
        "lev": "Leviticus",
        "lv": "Leviticus",
        "vayikra": "Leviticus",
        "numbers": "Numbers",
        "num": "Numbers",
        "nm": "Numbers",
        "bemidbar": "Numbers",
        "deuteronomy": "Deuteronomy",
        "deut": "Deuteronomy",
        "dt": "Deuteronomy",
        "devarim": "Deuteronomy",
        "joshua": "Joshua",
        "josh": "Joshua",
        "yehoshua": "Joshua",
        "judges": "Judges",
        "judg": "Judges",
        "jgs": "Judges",
        "shoftim": "Judges",
        "ruth": "Ruth",
        "ru": "Ruth",
        "1samuel": "1 Samuel",
        "1sam": "1 Samuel",
        "1sm": "1 Samuel",
        "1shmuel": "1 Samuel",
        "2samuel": "2 Samuel",
        "2sam": "2 Samuel",
        "2sm": "2 Samuel",
        "2shmuel": "2 Samuel",
        "1kings": "1 Kings",
        "1kgs": "1 Kings",
        "1melachim": "1 Kings",
        "2kings": "2 Kings",
        "2kgs": "2 Kings",
        "2melachim": "2 Kings",
        "1chronicles": "1 Chronicles",
        "1chron": "1 Chronicles",
        "1chr": "1 Chronicles",
        "2chronicles": "2 Chronicles",
        "2chron": "2 Chronicles",
        "2chr": "2 Chronicles",
        "ezra": "Ezra",
        "ezr": "Ezra",
        "nehemiah": "Nehemiah",
        "neh": "Nehemiah",
        "esther": "Esther",
        "est": "Esther",
        "job": "Job",
        "jb": "Job",
        "psalms": "Psalms",
        "psalm": "Psalms",
        "ps": "Psalms",
        "pss": "Psalms",
        "proverbs": "Proverbs",
        "prov": "Proverbs",
        "prv": "Proverbs",
        "ecclesiastes": "Ecclesiastes",
        "eccles": "Ecclesiastes",
        "eccl": "Ecclesiastes",
        "songofsolomon": "Song of Songs",
        "songofsongs": "Song of Songs",
        "songofsol": "Song of Songs",
        "sg": "Song of Songs",
        "isaiah": "Isaiah",
        "isa": "Isaiah",
        "yeshayahu": "Isaiah",
        "yeshaya": "Isaiah",
        "jeremiah": "Jeremiah",
        "jer": "Jeremiah",
        "yirmiyahu": "Jeremiah",
        "yirmiyah": "Jeremiah",
        "lamentations": "Lamentations",
        "lam": "Lamentations",
        "ezekiel": "Ezekiel",
        "ezek": "Ezekiel",
        "yechezkel": "Ezekiel",
        "daniel": "Daniel",
        "dan": "Daniel",
        "dn": "Daniel",
        "hosea": "Hosea",
        "hos": "Hosea",
        "hoshea": "Hosea",
        "joel": "Joel",
        "jl": "Joel",
        "yoel": "Joel",
        "amos": "Amos",
        "am": "Amos",
        "obadiah": "Obadiah",
        "obad": "Obadiah",
        "ob": "Obadiah",
        "ovadiah": "Obadiah",
        "ovadyah": "Obadiah",
        "jonah": "Jonah",
        "jon": "Jonah",
        "micah": "Micah",
        "mic": "Micah",
        "michah": "Micah",
        "nahum": "Nahum",
        "nah": "Nahum",
        "na": "Nahum",
        "nachum": "Nahum",
        "habakkuk": "Habakkuk",
        "hab": "Habakkuk",
        "hb": "Habakkuk",
        "chavakuk": "Habakkuk",
        "zephaniah": "Zephaniah",
        "zeph": "Zephaniah",
        "zep": "Zephaniah",
        "tzefaniah": "Zephaniah",
        "tzefanyah": "Zephaniah",
        "haggai": "Haggai",
        "hag": "Haggai",
        "hg": "Haggai",
        "chaggai": "Haggai",
        "zechariah": "Zechariah",
        "zech": "Zechariah",
        "zec": "Zechariah",
        "zecharya": "Zechariah",
        "zecharyah": "Zechariah",
        "zechariyah": "Zechariah",
        "malachi": "Malachi",
        "mal": "Malachi",
        # New Testament
        "matthew": "Matthew",
        "mathew": "Matthew",
        "matt": "Matthew",
        "mat": "Matthew",
        "mt": "Matthew",
        "mark": "Mark",
        "mk": "Mark",
        "luke": "Luke",
        "lk": "Luke",
        "john": "John",
        "jn": "John",
        "acts": "Acts",
        "actsoftheapostles": "Acts",
        "romans": "Romans",
        "rom": "Romans",
        "1corinthians": "1 Corinthians",
        "1cor": "1 Corinthians",
        "2corinthians": "2 Corinthians",
        "2cor": "2 Corinthians",
        "galatians": "Galatians",
        "gal": "Galatians",
        "ephesians": "Ephesians",
        "eph": "Ephesians",
        "philippians": "Philippians",
        "phil": "Philippians",
        "colossians": "Colossians",
        "col": "Colossians",
        "1thessalonians": "1 Thessalonians",
        "1thess": "1 Thessalonians",
        "1thes": "1 Thessalonians",
        "2thessalonians": "2 Thessalonians",
        "2thess": "2 Thessalonians",
        "2thes": "2 Thessalonians",
        "1timothy": "1 Timothy",
        "1tim": "1 Timothy",
        "1tm": "1 Timothy",
        "2timothy": "2 Timothy",
        "2tim": "2 Timothy",
        "2tm": "2 Timothy",
        "titus": "Titus",
        "ti": "Titus",
        "philemon": "Philemon",
        "philem": "Philemon",
        "phlm": "Philemon",
        "hebrews": "Hebrews",
        "heb": "Hebrews",
        "james": "James",
        "jas": "James",
        "1peter": "1 Peter",
        "1pet": "1 Peter",
        "1pt": "1 Peter",
        "2peter": "2 Peter",
        "2pet": "2 Peter",
        "2pt": "2 Peter",
        "1john": "1 John",
        "1jn": "1 John",
        "2john": "2 John",
        "2jn": "2 John",
        "3john": "3 John",
        "3jn": "3 John",
        "jude": "Jude",
        "revelation": "Revelation",
        "revelations": "Revelation",
        "rev": "Revelation",
        "rv": "Revelation",
        # Deuterocanon
        "judith": "Judith",
        "judeth": "Judith",
        "jdt": "Judith",
        "wisdom": "Wisdom",
        "wis": "Wisdom",
        "wisdomofsolomon": "Wisdom",
        "tobit": "Tobit",
        "tob": "Tobit",
        "sirach": "Ecclesiasticus",
        "sir": "Ecclesiasticus",
        "ecclesiasticus": "Ecclesiasticus",
        "baruch": "Baruch",
        "bar": "Baruch",
        "1maccabees": "1 Maccabees",
        "1macc": "1 Maccabees",
        "1mac": "1 Maccabees",
        "2maccabees": "2 Maccabees",
        "2macc": "2 Maccabees",
        "2mac": "2 Maccabees",
        "3maccabees": "3 Maccabees",
        "3macc": "3 Maccabees",
        "3mac": "3 Maccabees",
        "4maccabees": "4 Maccabees",
        "4macc": "4 Maccabees",
        "4mac": "4 Maccabees",
        "restofdaniel": "Prayer of Azariah",
        "additionstodaniel": "Prayer of Azariah",
        "adddan": "Prayer of Azariah",
        "songofthethreechildren": "Prayer of Azariah",
        "prayerofazariah": "Prayer of Azariah",
        "restofesther": "Additions to Esther",
        "additionstoesther": "Additions to Esther",
        "addesth": "Additions to Esther",
        "prayerofmanasses": "Prayer of Manasseh",
        "prayerofmanasseh": "Prayer of Manasseh",
        "manasses": "Prayer of Manasseh",
        "manasseh": "Prayer of Manasseh",
        "prman": "Prayer of Manasseh",
        "1esdras": "1 Esdras",
        "1esd": "1 Esdras",
        "2esdras": "2 Esdras",
        "2esd": "2 Esdras",
        "storyofsusanna": "Susanna",
        "susanna": "Susanna",
        "sus": "Susanna",
        "belandthedragon": "Bel and the Dragon",
        "bel": "Bel and the Dragon"
    }

    try:
        converted_name = books[name.lower().replace(" ", "")]
        return converted_name
    except KeyError:
        return None


def get_book_number(book_name):
    """ Find appropriate book number for book name. This is used in
    situations where local bibles are used, or to determine if a translation
    supports a book (mostly the deuterocanon).

    :param book_name: The book to get the book number for
    """
    book_numbers = {
        "Genesis": 1,
        "Exodus": 2,
        "Leviticus": 3,
        "Numbers": 4,
        "Deuteronomy": 5,
        "Joshua": 6,
        "Judges": 7,
        "Ruth": 8,
        "1 Samuel": 9,
        "2 Samuel": 10,
        "1 Kings": 11,
        "2 Kings": 12,
        "1 Chronicles": 13,
        "2 Chronicles": 14,
        "Ezra": 15,
        "Nehemiah": 16,
        "Esther": 17,
        "Job": 18,
        "Psalms": 19,
        "Proverbs": 20,
        "Ecclesiastes": 21,
        "Song of Songs": 22,
        "Isaiah": 23,
        "Jeremiah": 24,
        "Lamentations": 25,
        "Ezekiel": 26,
        "Daniel": 27,
        "Hosea": 28,
        "Joel": 29,
        "Amos": 30,
        "Obadiah": 31,
        "Jonah": 32,
        "Micah": 33,
        "Nahum": 34,
        "Habakkuk": 35,
        "Zephaniah": 36,
        "Haggai": 37,
        "Zechariah": 38,
        "Malachi": 39,
        "Matthew": 40,
        "Mark": 41,
        "Luke": 42,
        "John": 43,
        "Acts": 44,
        "Romans": 45,
        "1 Corinthians": 46,
        "2 Corinthians": 47,
        "Galatians": 48,
        "Ephesians": 49,
        "Philippians": 50,
        "Colossians": 51,
        "1 Thessalonians": 52,
        "2 Thessalonians": 53,
        "1 Timothy": 54,
        "2 Timothy": 55,
        "Titus": 56,
        "Philemon": 57,
        "Hebrews": 58,
        "James": 59,
        "1 Peter": 60,
        "2 Peter": 61,
        "1 John": 62,
        "2 John": 63,
        "3 John": 64,
        "Jude": 65,
        "Revelation": 66,
        "Judith": 67,
        "Wisdom": 68,
        "Tobit": 69,
        "Ecclesiasticus": 70,
        "Baruch": 71,
        "1 Maccabees": 72,
        "2 Maccabees": 73,
        "3 Maccabees": 74,
        "4 Maccabees": 75,
        "Prayer of Azariah": 76,
        "Additions to Esther": 77,
        "Prayer of Manasseh": 78,
        "1 Esdras": 79,
        "2 Esdras": 80,
        "Susanna": 81,
        "Bel and the Dragon": 82
    }

    try:
        book_number = book_numbers[book_name]
        return book_number
    except KeyError:
        return None
